client/lib/subscription.dart

```diff
@@ -18,82 +24,139 @@
-              itemRemoved = _checkIsBroadCast(itemRemoved),
-              _closed = false;
+    StateSubscription._(Future<pb.Subscription> subscription,
+                        Stream<T> itemAdded,
+                        Stream<T> itemChanged,
+                        Stream<T> itemRemoved):
+            itemChanged = _checkBroadcast(itemChanged),
+            itemRemoved = _checkBroadcast(itemRemoved),
+            _closed = false,
+            super._(subscription, itemAdded);
 
     /// Creates a new instance which broadcasts updates from under the given Firebase node.
-    factory Subscription.of(FirebaseSubscription firebaseSubscription,
-                            FirebaseClient database) {
-        var subscription = firebaseSubscription.subscription;
-        var typeUrl = subscription.topic.target.type;
-        var builderInfo = theKnownTypes.findBuilderInfo(typeUrl);
-        if (builderInfo == null) {
-            throw ArgumentError.value(firebaseSubscription, 'firebase subscription',
-                                      'Firebase subscription type `${typeUrl} is unknown.');
-        }
-        var nodePath = firebaseSubscription.nodePath.value;
-
-        var itemAdded = database
-            .childAdded(nodePath)
-            .map((json) => parseIntoNewInstance<T>(builderInfo, json));
-        var itemChanged = database
-            .childChanged(nodePath)
-            .map((json) => parseIntoNewInstance<T>(builderInfo, json));
-        var itemRemoved = database
-            .childRemoved(nodePath)
-            .map((json) => parseIntoNewInstance<T>(builderInfo, json));
-
-        return Subscription(subscription, itemAdded, itemChanged, itemRemoved);
+    factory StateSubscription.of(Future<FirebaseSubscription> firebaseSubscription,
+                                 BuilderInfo builderInfoForType,
+                                 FirebaseClient database) {
+        var subscription = firebaseSubscription.then((value) => value.subscription);
+        var nodePath = _nodePath(firebaseSubscription);
+        var itemAdded = _nodePathStream(nodePath)
+            .asyncExpand((element) => database.childAdded(element))
+            .map((json) => parseIntoNewInstance<T>(builderInfoForType, json));
+        var itemChanged = _nodePathStream(nodePath)
+            .asyncExpand((element) => database.childChanged(element))
+            .map((json) => parseIntoNewInstance<T>(builderInfoForType, json));
+        var itemRemoved = _nodePathStream(nodePath)
+            .asyncExpand((element) => database.childRemoved(element))
+            .map((json) => parseIntoNewInstance<T>(builderInfoForType, json));
+        return StateSubscription._(subscription, itemAdded, itemChanged, itemRemoved);
     }
+}
 
-    bool get closed => _closed;
+/// A subscription for events.
+///
+/// To consume unpacked typed event messages, use [eventMessages]. To use events with metadata,
+/// use [events].
```

Comment:
I wouldn't let end-users to consume `Event`s. Let's discuss this later.

Reply:
As discussed, for consistency with other clients, as well as for the needs of users who may build logic on `EventContext`s, we'll support both events and event messages.